crates/circuit/src/parameter/parameter_expression.rs

```diff
@@ -0,0 +1,2162 @@
+        &self,
+        map: &HashMap<Symbol, Value>,
+        allow_unknown_parameters: bool,
+    ) -> Result<Self, ParameterError> {
+        // The set of symbols we will bind. Used twice, hence pre-computed here.
+        let bind_symbols: HashSet<Symbol> = map.keys().cloned().collect();
+
+        if !allow_unknown_parameters {
+            let existing_symbols: HashSet<Symbol> = self.name_map.values().cloned().collect();
+            let difference: HashSet<Symbol> = bind_symbols
+                .difference(&existing_symbols)
+                .cloned()
+                .collect();
+            if !difference.is_empty() {
+                return Err(ParameterError::UnknownParameters(difference));
+            }
+        }
+
+        // bind the symbol expression and then check the outcome for inf/nan, or numeric values
+        let bound_expr = self.expr.bind(map);
+        let bound = match bound_expr.eval(true) {
+            Some(v) => match &v {
+                Value::Real(r) => {
+                    if r.is_infinite() {
+                        Err(ParameterError::BindingInf)
+                    } else if r.is_nan() {
+                        Err(ParameterError::BindingNaN)
+                    } else {
+                        Ok(SymbolExpr::Value(v))
+                    }
+                }
+                Value::Int(_) => Ok(SymbolExpr::Value(v)),
+                Value::Complex(c) => {
+                    if c.re.is_infinite() || c.im.is_infinite() {
+                        Err(ParameterError::ZeroDivisionError) // TODO this should probs be BindingInf
+                    } else if c.re.is_nan() || c.im.is_nan() {
+                        Err(ParameterError::BindingNaN)
+                    } else if (-symbol_expr::SYMEXPR_EPSILON..symbol_expr::SYMEXPR_EPSILON)
+                        .contains(&c.im)
+                    {
+                        Ok(SymbolExpr::Value(Value::Real(c.re)))
+                    } else {
+                        Ok(SymbolExpr::Value(v))
+                    }
+                }
+            },
+            None => Ok(bound_expr),
+        }?;
+
+        // update the name map by removing the bound parameters
+        let bound_name_map: HashMap<String, Symbol> = self
+            .name_map
+            .iter()
+            .filter(|(_, symbol)| !bind_symbols.contains(*symbol))
+            .map(|(name, symbol)| (name.clone(), symbol.clone()))
+            .collect();
+
+        Ok(Self {
+            expr: bound,
+            name_map: bound_name_map,
+        })
+    }
+
+    /// Merge name maps.
+    ///
+    /// # Arguments
+    ///
+    /// * `other` - The other parameter expression whose symbols we add to self.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(HashMap<String, Symbol>)` - The merged name map.
+    /// * `Err(ParameterError)` - An error if there was a name conflict.
+    fn update_name_map(&self, other: &Self) -> Result<HashMap<String, Symbol>, ParameterError> {
```

Comment:
```suggestion
    fn merged_name_map(&self, other: &Self) -> Result<HashMap<String, Symbol>, ParameterError> {
```

Otherwise this is confusing unless you click in to see that it accepts `&self` not `&mut self`.

Reply:
also the doc says "merge name maps" lol